featurizers/map4_mhfp_helper_functions.py

```diff
@@ -0,0 +1,140 @@
+    atom_map = {}
+
+    submol = PathToSubmol(Mol, env, atomMap=atom_map)
+
+    if idx in atom_map:
+        smiles = MolToSmiles(
+            submol,
+            rootedAtAtom=atom_map[idx],
+            canonical=True,
+            isomericSmiles=False,
+        )
+        return smiles
+    else:
+        return ""
+
+
+def _get_atom_envs(Mol: Mol, radius: int):
+    """
+    For each atom get its environment.
+    """
+    atoms_env = defaultdict(lambda: [])
+    for atom in Mol.GetAtoms():
+        idx = atom.GetIdx()
+        for r in range(1, radius + 1):
+            atoms_env[idx].append(_find_neighborhood(Mol, idx, r))
+
+    return atoms_env
+
+
+def _all_pairs(Mol: Mol, atoms_envs: dict, radius: int, is_counted: bool):
+    """
+    Gets a list of atom-pair molecular shingles - circular structures written as SMILES, separated by the bond distance
+    between the two atoms along the shortest path.
+    """
+    atom_pairs = []
+    distance_matrix = GetDistanceMatrix(Mol)
+    num_atoms = Mol.GetNumAtoms()
+    shingle_dict = defaultdict(int)
+
+    # Iterate through all pairs of atoms and radius. Shingles are stored in format:
+    # (neighborhood of atom A of radius i) | (distance between atom A and atom B) | (neighborhood of atom B of radius i)
+    #
+    # If we want to count the shingles (is_counted), we increment their value in shingle_dict.
+    # After nested for-loop, all shingles, with their respective counts, will be added to atom_pairs list.
+    for idx_1, idx_2 in itertools.combinations(range(num_atoms), 2):
+        # distance_matrix consists of floats as integers, so they need to be converted to integers first
+        dist = str(int(distance_matrix[idx_1][idx_2]))
+
+        for i in range(radius):
+            env_a = atoms_envs[idx_1][i]
+            env_b = atoms_envs[idx_2][i]
+
+            if not len(env_a) or not len(env_b):
+                continue
+
+            ordered = sorted([env_a, env_b])
+
+            shingle = f"{ordered[0]}|{dist}|{ordered[1]}"
+
+            if is_counted:
+                shingle_dict[shingle] += 1
+            else:
+                atom_pairs.append(shingle.encode("utf-8"))
+
+    if is_counted:
+        # Shingle in a format:
+        # (neighborhood of atom A of radius i) | (distance between atom A and atom B) | \
+        # (neighborhood of atom B of radius i) | (shingle count)
+        for shingle, shingle_val in shingle_dict.items():
+            shingle += "|" + str(shingle_val)
+            atom_pairs.append(shingle.encode("utf-8"))
+
+    return atom_pairs
+
+
+def get_map4_fingerprint(
+    Mol: Mol,
+    dimensions: int,
+    radius: int,
+    is_counted: bool,
+    random_state: int,
+):
+    # TODO - There are certain molecules, for which this function will return a error:
+    #   https://github.com/Arch4ngel21/emf/issues/13
+    #   So for now it's handled by try/except
+    try:
+        from datasketch.minhash import MinHash
```

Comment:
We should inform user that `datasketch` is required. Not necessarily in this PR, but add issue for better handling this